examples/threading/threading.c

```diff
@@ -0,0 +1,90 @@
+#include "threading.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+// Optional: use these functions to add debug or error prints to your application
+#define DEBUG_LOG(msg,...)
+//#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
+#define ERROR_LOG(msg,...) printf("threading ERROR: " msg "\n" , ##__VA_ARGS__)
+
+
+
+void* threadfunc(void* thread_param)
+{
+
+    // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
+    // hint: use a cast like the one below to obtain thread arguments from your parameter
+    //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+     struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+
+    // Initialize thread_complete_success to false
+    thread_func_args->thread_complete_success = false;
+
+    // Wait before attempting to obtain the mutex
+    usleep(thread_func_args->wait_to_obtain_ms * 1000); // Convert milliseconds to microseconds
+
+    // Try to lock the mutex
+    if (pthread_mutex_lock(thread_func_args->mutex) != 0) {
+        ERROR_LOG("Failed to obtain mutex\n");
+        // Free thread data and return NULL to indicate an error occurred
+        free(thread_param);
+        return NULL;
+    }
+
+    DEBUG_LOG("Mutex obtained by thread\n");
+
+    // Wait before releasing the mutex
+    usleep(thread_func_args->wait_to_release_ms * 1000); // Convert milliseconds to microseconds
+
+    // Unlock the mutex
+    if (pthread_mutex_unlock(thread_func_args->mutex) != 0) {
+        ERROR_LOG("Failed to release mutex\n");
+        // Free thread data and return NULL to indicate an error occurred
+        free(thread_param);
```

Comment:
@shruthi-thallapally I don't think you should free thread_param here, I think this will still be accessed by the thread join to query for completion status.